verification/spec/connection-handshake/Environment.tla

```diff
@@ -0,0 +1,233 @@
+
+
+chmA == INSTANCE ConnectionHandshakeModule
+        WITH MaxHeight      <- MaxHeight,
+             inBuf          <- bufChainA,
+             outBuf         <- bufChainB,
+             store          <- storeChainA,
+             ConnectionIDs  <- chainAParameters.connectionID,
+             ClientIDs      <- chainAParameters.clientID
+
+
+chmB == INSTANCE ConnectionHandshakeModule
+        WITH MaxHeight      <- MaxHeight,
+             inBuf          <- bufChainB,      \* Flip the message buffers w.r.t. chain A buffers.
+             outBuf         <- bufChainA,      \* Inbound for "A" is outbound for "B".
+             store          <- storeChainB,
+             ConnectionIDs  <- chainBParameters.connectionID,
+             ClientIDs      <- chainBParameters.clientID
+
+
+ConnectionStates == {"UNINIT", "INIT", "TRYOPEN", "OPEN"}
+
+
+ConnectionParameters ==
+    [
+        localEnd : chmA!ConnectionEnds,
+        remoteEnd : chmB!ConnectionEnds
+    ]
+
+Connections ==
+    [
+        parameters : ConnectionParameters,
+        state : ConnectionStates
+    ]
+
+Proofs ==
+    [
+        height : 1..MaxHeight
+    ]
+
+(******************************** Messages ********************************
+ These messages are connection handshake specific.
+ 
+ In the low-level connection handshake protocol, the four messages have the
+ following types: ConnOpenInit, ConnOpenTry, ConnOpenAck, ConnOpenConfirm.
+ These are described in ICS 003.
+ In this high-level specification, we choose slightly different names, to
+ make an explicit distinction to the low-level protocol. Message types
+ are as follows: CHMsgInit, CHMsgTry, CHMsgAck, and CHMsgConfirm. Notice that
+ the fields of each message are also different to the ICS 003 specification.
+ ***************************************************************************)
+ConnectionHandshakeMessages ==
+    [type : {"CHMsgInit"}, 
+     parameters : ConnectionParameters]
+    \union
+    [type : {"CHMsgTry"},
+     parameters : ConnectionParameters,
+     stateProof : Proofs,
+     clientProof : Proofs]
+    \union
+    [type : {"CHMsgAck"},
+     parameters : ConnectionParameters,
+     stateProof : Proofs,
+     clientProof : Proofs]
+     \union
+    [type : {"CHMsgConfirm"},
+     parameters : ConnectionParameters,
+     stateProof : Proofs]
+
+
+(* The set of all Init messages, such that the local end is the
+    set 'le', and the remote end is set 're'. *)
+InitMsgs(le, re) ==
+    [type : {"CHMsgInit"},
+     parameters : [localEnd : le,
+                   remoteEnd : re]]
+
+
+(* TODO: find a more elegant way to do this. 
+    What we want, ideally, is to specify this:
+        z \in x
+    to obtain a random Sequence z out of a set of sequences x, obtaining:
+        bufChainA \in InitMsgs
+    instead of the awkward lines for building bufChainA and bufChainB below.
+ *)
+ChooseInitMsg(le, re) ==
+    CHOOSE x \in InitMsgs(le, re) : TRUE
```

Comment:
You normally don't want to do this as choose will deterministically pick one. This constrain model checker in a bad way. You probably want here to pick message from the set of all InitMsgs, for example \E m \in InitMsgs.